### ISSUE TITLE
RFE: Implement Service Monitoring logic

### ISSUE THREAD

**Issue by mrguitar** [systemd/systemd#9075]
**Is your feature request related to a problem? Please describe.**

systemd provides service resiliency with options like Restart=, but it currently doesn't offer the means to ensure the service is running in a proper state. Other systems like Kuberenetes provide a simple means to run basic (or custom) health checks and it would be powerful if systemd adopted something similar. 

**Describe the solution you'd like**
We discussed adding something like ExecMonitor= in the unit so users can define a shell script running in a loop to monitor the service. If the loop fails, the Restart= policy would be run. This concept would improve the resiliency for services, systems, devices, etc and further eliminate the need for certain single-node clustering logic.

**Comment by keszybz** [comment]
This has come up before. I think this should be implemented as a separate unit. Systemd already implements the general checks whether the pid is alive and sending watchdog pings. Additional checks on top would require some custom logic and configuration. This means that the monitor can nicely be implemented as a service unit (possibly a template unit to monitor a certain class of service units). This way we avoid the complicated semantics of what happens when the monitor service dies but the main process is still alive, etc. The monitor service can also have its own user, be less privileged, have much stricter resource constraints. It is also possible to have multiple monitor services, and start and stop them at will. If the monitor service detects that the monitoree is failing, it should just send normal systemd commands to restart the unit. Would this work for you? Is there any functionality missing to make this easy?

**Comment by korekhov** [comment]
IMHO, an extra service would bring more complications. It is basically the same that I have to do today - cronjobs that run every minute and perform a health check on a service they monitor and send systemd commands. This obviously has a number of issues - crond may die, cronjobs may get deleted, service could be down for longer because of 1 minute intervals, etc.

Imagine nginx or Apache running on your systemd-managed OS. It watches the process, but that does not mean the process is actually providing a proper service - for example, it may be giving 500 errors to any HTTP request that comes in for whatever reason. In situation like this, it may more desirable to shut down such instance after a number of auto-restarts instead of keep giving the errors to the clients. Currently systemd unit does not provide an ability to implement that thus I made this request/suggestion to Ben last year. And to be honest, I don't think this warrants a separate service. This "in-depth" monitoring (in above example could be just a curl call to read the HTTP code from a service) to me feels like a part of an nginx/apache systemd unit. This looks like the simplest scenario to me from systemd user perspective. I'm sure it adds more complications on systemd side though.

**Comment by keszybz** [comment]
I don't see anything in the description which would make it better to make it necessary or even good to make the monitoring tool part of the original tool. Under systemd, services are commonly split into multiple units, because that provides flexibility and clarity. For example, a .timer + .socket + .service combo  to implement something that was a single inet service before is rather common. That can be annoying for simple things, but that's how things are with systemd and avoiding this is a very weak argument.

The monitoring unit can have `[Install] RequiredBy=httpd.service`, so that when it httpd.service is started, the monitoring service is guaranteed to be started too.

In fact, the monitoring is **better** if it is separate. For example, network namespaces or selinux could create a situation where the service "works" internally, but is not accessible for any other process. By making the monitoring service separate, we check things more realistically.

Also, nowadays we try to minimize privileges. Making the monitoring unit separate allows it to be very strongly sandboxed, and avoids inheriting the (fairly extensive) privileges of the apache service.

I think we should try to come up with a recipe for implementing this using a separate unit, and getting the dependencies right (e.g. I think `PartOf=` should be used), and maybe extend systemd if any bits are missing to make this convenient.

**Comment by korekhov** [comment]
Ah, I guess I misunderstood you at first - yes, if this is going to result in <app>**.monitor** unit, then it is perfectly fine. I thought you were proposing a single <global_monitor>.service instead.

How do you envision the actual monitoring task is going to executed by .monitor unit? Some kind of loop or execution at some defined frequency (every 5 sec, etc.)? If you take "curl" above as an example again, how would I run it from this unit?

**Comment by jritter** [comment]
I would love to see such a feature in systemd. I think a possible solution could be the following:
* myservice.service file with Type=notify
* myservice.monitor which
  * periodically triggers a monitoring command towards the service behind myservice.service, for instance a curl to a http service
  * based on the return value of the monitoring command it sends sd_notify watchdog and ready notification for myservice.service

This would nicely integrate with the already existing readyness and watchdog mechanisms.

While I think the sd_notify mechanism is great, it is somewhat limited to services which support this feature. Java applications for instance typically don't implement sd_notify.

Any thoughts?

**Comment by korekhov** [comment]
Basically, all what I was looking for at that time is what Kubernetes has implemented as probes - https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/
Ideally, systemd has it all in a single manifest rather than a separate type of service, but implementation details are just details. The main thing here is allow user to monitor not just a process, but actually be sure that if it is running, it is giving expected results and not just a process that gives its users some errors.
Thanks!
Konstantin Orekhov 

    On Wednesday, January 6, 2021, 01:52:55 PM PST, Juerg Ritter <notifications@github.com> wrote:  
 
 


I would love to see such a feature in systemd. I think a possible solution could be the following:
   
   - myservice.service file with Type=notify
   - myservice.monitor which      
      - periodically triggers a monitoring command towards the service behind myservice.service, for instance a curl to a http service
      - based on the return value of the monitoring command it sends sd_notify watchdog and ready notification for myservice.service


This would nicely integrate with the already existing readyness and watchdog mechanisms.

While I think the sd_notify mechanism is great, it is somewhat limited to services which support this feature. Java applications for instance typically don't implement sd_notify.

Any thoughts?

**Comment by jritter** [comment]
I agree @korekhov, IMHO the goal should be something like Kubernetes liveness/readiness probes. The question whether these probes should be configured in the service unit file, or in a separate unit file is secondary. Personally I think it would be cleaner and more stringent with current architecture to keep this in a separate unit.

**Comment by boucman** [comment]
with NotifyAccess you can give any process in the service the permission to use sd_notify

At that point a process can spawn its own monitoring process and let that process do the monitoring

would that work for you ?

**Comment by jritter** [comment]
Yeah, I have seen this feature, IMHO it is not a very elegant solutions when I'm not under the control of the application which I'm trying to orchestrate, it would involve wrapper scripts which have to be replicated for every service. I would love to have a well defined and easy to use docking point for such probes, the same way Kubernetes handles it for instance.

**Comment by boucman** [comment]
I don't know.. I don't fine spawning your own monitoring process inelegant... it allows much more flexibility since you can do funky things with fork/exec... a wrapper script is not perfect, but it does give you the best control on what's going on and since the monitoring application is also under the same unit, systemd will properly kill/restart it with your main application. It's probably more robust from a monitoring PoV

I mean... adding an ExecMonitor= command that would launch a second application in a similar env to the main one, but which would have the notify socket and not other stuff (pre-opened sockets ?) might make sense, but this probably needs a more precise spec (when it is started ? probably before ExecStartPost, how is it started ? in what environment ? how is it monitored ?)

I'm not sure that adding that to systemd instead of having a simple wrapper makes much sense. maybe for socket-based applications ?

**Comment by jritter** [comment]
I agree with you that spawning your own process offers the most flexibility. But it is a tradeoff with complexity from a user perspective. Systemd takes away so much complexity and makes it easy for the user to define its own services, and I think this is part of the reason why systemd has been adopted so widely. 

I am a system engineer and not a software engineer, and I don't think I have the whole view of all ramifications here, but I guess part of such a discussion is to discuss possible solutions and take it from there.

So here is my high level view on how a service definition of such a mechanism could look like, and I lean heavily on the Kubernetes approach (https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/) which is IMHO very simple to use from a user perspective and also works really well.

```
[Service]
Restart=on-failure
ExecStart=/path/to/my/simple/service -p 3000
Type=notify
ReadinessProbeCommand=/usr/bin/curl http://localhost:3000 --fail --silent
ReadinessInitialDelay=20s
ReadinessPeriod=5s
ReadinessFailureThreshold=1
LivenessProbeCommand=/usr/bin/curl http://localhost:3000 --fail --silent
LivenessInitialDelay=20s
LivenessPeriod=5s
LivenessFailureThreshold=5
```

Some remarks:

- Systemd would periodically trigger the specified commands for liveness and readyness probe (in Kubernetes this is done by the Kubelet)
- If the ReadinessProbeCommand returns 0, READY=1 would be sent through sd_notify --> service is ready
- LivenessProbes and Readyness Probes update watchdog (WATCHDOG=1) through sd_notify if command executes successfully
- Other than that, all the properties of Type=notify would apply (restart policies, watchdog policies, etc.)
- Maybe it could even make sense to put these Probe definitons in a separate unit file with extension .monitor or .probe, sharing the socket or timer philosophy

I'm looking forward continuing this discussion!

**Comment by boucman** [comment]
Readiness already work, you do a Type=simple and an ExecStartPost=<your probing command in a loop>

since systemd will wait for ExecStartPost to finish before declaring your service ready, this will block readiness
And no... services should be self-contained. readiness/liveness should definitely be part of the service.

sockets and timers are a about triggering the unit, i.e separating"when is the unit started" and "how is the unit started"

**Comment by Werkov** [comment]
The systemd entities are (mainly) services which are opaque from PID 1 perspective (if a process runs, then it's fine; if not, the service can be (re)started anew). The notifications allow passing some finer information to PID 1 about the service but the applicable means remain the same (basically summed via `Restart=`). I see there is a mismatch between PID 1 low-level operations and such high-level checks. How could PID 1 interpret results of such high-level checks? I expect their failures would typically require an action out of PID 1.

**Comment by jritter** [comment]
> Readiness already work, you do a Type=simple and an ExecStartPost=
> 
> since systemd will wait for ExecStartPost to finish before declaring your service ready, this will block readiness
> And no... services should be self-contained. readiness/liveness should definitely be part of the service.
> 
> sockets and timers are a about triggering the unit, i.e separating"when is the unit started" and "how is the unit started"

I fully agree with this, the primary use case I'm trying to find a solution for is modeling service dependenies, where I would like to rely on the systemd service status, which ideally reflects the readiness of the service, rather than the existence of the process.

**Comment by jritter** [comment]
> The systemd entities are (mainly) services which are opaque from PID 1 perspective (if a process runs, then it's fine; if not, the service can be (re)started anew). The notifications allow passing some finer information to PID 1 about the service but the applicable means remain the same (basically summed via `Restart=`). I see there is a mismatch between PID 1 low-level operations and such high-level checks. How could PID 1 interpret results of such high-level checks? I expect their failures would typically require an action out of PID 1.

My idea would be to post sd_notify notifications (READY=1 and WATCHDOG=1) on behalf of the application if the application reports healthy, based on the result of the command. If I understand systemd correctly, this could be done outside of PID 1, we just need a docking point in systemd from where we can trigger this execution.

Restarts would be done according to the already existing restart policy, so IMHO in this area there wouldn't be changes necessary.

**Comment by Werkov** [comment]
Ad dependencies -- are we still talking about the HTTP endpoint example? If that's a dependency for another service, then (already existing) socket activation looks like the way to go. Is the information yielded from the probe to be interpreted on the machine or by an external entity?

Ad docking point -- I'm afraid it's not enough, as noted by @boucman above, you'd also need to define lifecycle of such a probe and it's interaction with the lifecycle of the service.  I meant in the previous comment that the high-level check can fail e.g. because of service misconfiguration, running out of resources or similar, which can't be properly fixed by mere restarting of the service, you need a higher action (or am I misunderstading the goal?)

**Comment by jritter** [comment]
> Ad dependencies -- are we still talking about the HTTP endpoint example? If that's a dependency for another service, then (already existing) socket activation looks like the way to go. Is the information yielded from the probe to be interpreted on the machine or by an external entity?

Socket activation is great, the caveat in my case is that the service I'm trying to integrate is not capable to activate itself through socket activation. 

Maybe I can outline my situation a bit more clear: I'm a system engineer assigned to a project which deals with a information kiosk solution. The kiosk app is a spring boot application, which runs locally on the kiosk system, and the users can interact with it through a browser in kiosk mode. I get these finished components of software on my table, and my task is to reliably integrate it into the service architecture of RHEL (systemd). For starting up, I have this hard dependency: The browser can never start before the spring boot application is started and **ready to serve requests**. The spring boot application is neither aware of sd_notify nor socket activation, but it exposes a HTTP health endpoint which can be used to query the health of the spring boot application.

Ideally I would like to model this in two systemd services, one that handles the browser, and the other for the spring boot service. But the browser can only start when the spring boot application is up and ready to serve requests.

> 
> Ad docking point -- I'm afraid it's not enough, as noted by @boucman above, you'd also need to define lifecycle of such a probe and it's interaction with the lifecycle of the service. I meant in the previous comment that the high-level check can fail e.g. because of service misconfiguration, running out of resources or similar, which can't be properly fixed by mere restarting of the service, you need a higher action (or am I misunderstading the goal?)

With "docking point" I was thinking about an option in the Service section (such as ExecStartPre or ExecStartPost) where I can define a command which periodically fires (for instance curl which queries a HTTP endpoint), and then posts sd_notify notifications (READY=1 and WATCHDOG=1 if exit code 0, nothing otherwise) on behalf of the sd_notify unaware service. 

My take on service misconfigurations: If you opt to use such a feature it is your responsibility to wire up everything correctly. It's the same in Kubernetes for instance, if readiness/lifeness probes are misconfigured, pods keep restarting and will never be ready.

I am fully aware that this scenario is not 100% in line with systemd best practices, but the truth is that I keep running in these scenarios, and I also got questions multiple times how this can be solved best. In many situations, a system engineer has to work with what's on his/her table (for example a Spring Boot application) and has to make it work. I've been looking for a while now, but I haven't found a nice solution. My definition of a solution would be a spring boot application that exposes a health endpoint and a systemd service which would switch to active only when the application healthcheck reports healthy. The watchdog portion would be a nice extra I suppose.

**Comment by boucman** [comment]
for your readiness problem, add an ExecStartPost to your spring part that blocks until the service is ready (a shell script that loop-attempts to connect to the monitor URL will do)
* The spring service will be __starting__ until ExecStartPost return
* The startup will be protected by a systemd managed timeout, so you wont stay stuck 
After that, the browser service can just use After= to be correctly synchronized with the spring service

The usuall "design pattern" is to do config file checking in an ExecStartPre that will fail if the configuration is incorrect.

As for liveliness probe, it could be done with timers/services that would stop other services, but I really think this should be internal to a service and not systemd's job to manage... I'm not a systemd developer so my opinion doesn't really matter, but I don't think the way you are proposing to do stuff would fly with the systemd designers...

maybe Some sort of WatchdogProbe= that would be started when WatchdogSec is reached and could have a chance to "save" the service by sending a readiness signal ? I don't know...
